### PR TITLE
Implement Check for Storage Space for IoT Edge

### PR DESCRIPTION
**Pull request by nimanch**
This PR adds a check to make sure the target device has minimum required storage space for installing azure iot edge binaries and runtime containers.  

1. Adds the check for Storage Space
2. Adds placeholders for storage binaries for each of the supported architectures. 
3. Refactors the Shared Library Check. 
4. Some minor formatting fixes

Todo: 
1. For every release, automatically make sure that the binary size and container size of iotedge match what's present in the script
2. Update numbers for AMD64 and ARM64

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [ ] concise summary of tests added/modified
	- [ ] local testing done.  

### Draft PRs
- Open the PR in `Draft` mode if it is:
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.

_Note: We use the kodiakhq bot to merge PRs once the necessary checks and approvals are in place. When it merges a PR, kodiakhq converts the PR title to the commit title, PR description to the commit description, and squashes all the commits in the PR to a single commit. The net effect is that entire PR becomes a single commit. Please follow the best practices mentioned [here](https://chris.beams.io/posts/git-commit/#:~:text=The%20seven%20rules%20of%20a%20great%20Git%20commit,what%20and%20why%20vs.%20how%20For%20example%3A%20) for the PR title and description_

### CODE REVIEW

File: platform-validation/scripts/aziot-compatibility.sh [hunk 1]
```diff
@@ -501,58 +547,71 @@ check_shared_library_dependency() {
...
+x86_64_iotedge_binaries_size=42.39
+x86_64_iotedge_binaries_avg_memory=54.24
+x86_64_iotedge_container_size=254.96
+x86_64_iotedge_container_memory=245.5
+aarch64_iotedge_binaries_size=36.68
+aarch64_iotedge_binaries_avg_memory=26.62
+aarch64_iotedge_container_size=322.6
+aarch64_iotedge_container_memory=154.53
+iot_edge_size_buffer=50
+iot_edge_memory_buffer=50
```
Comment (damonbarry) [inline platform-validation/scripts/aziot-compatibility.sh:614]: Are `iot_edge_size_buffer` and `iot_edge_memory_buffer` used anywhere?

File: platform-validation/scripts/aziot-compatibility.sh [hunk 2]
```diff
@@ -501,58 +547,71 @@ check_shared_library_dependency() {
...
+    application_size=$2
+
+    # Check dependencies
+    if ! need_cmd df; then
+        wrap_warn "check_storage_space"
+        wrap_warning "Could not find df utility to calculate disk space, Skipping the check"
+        return
+    fi
+
+    if ! echo "$application_size" | grep -q '^[0-9]*.[0-9]*$'; then
```
[the remark this answers is not in the document]
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: I fixed the regex so that for any non-number, we simply throw an error

File: platform-validation/scripts/aziot-compatibility.sh [hunk 3]
```diff
@@ -576,11 +577,13 @@ check_storage_space() {
         return
     fi
 
-    if ! echo "$application_size" | grep -q '^[0-9]*.[0-9]*$'; then
+    if ! echo "$application_size" | grep -q '^[.0-9]*$'; then
```
Comment (damonbarry) [inline platform-validation/scripts/aziot-compatibility.sh:580]: This new regex still lets empty string pass through without an error.
```bash
$ echo '' | grep -q '^[.0-9]*$'
$ echo $?
0
```

File: platform-validation/scripts/aziot-compatibility.sh [hunk 4]
```diff
@@ -576,11 +577,13 @@ check_storage_space() {
...
         wrap_bad "Invalid Application Size provided" "$application_size"
         return 2
     fi
 
+    #Provide a Buffer of 50MB to account for any additional log files and storage.
```
Comment (damonbarry) [inline platform-validation/scripts/aziot-compatibility.sh:585]: Nit: This comment shouldn't refer to the `$buffer` variable's underlying value (50MB)

File: platform-validation/scripts/aziot-compatibility.sh (hunk 3 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: I am having a very bad regex day :D, I pushed another change, and tested the empty case as well

File: platform-validation/scripts/aziot-compatibility.sh (hunk 4 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: Fixed

File: platform-validation/scripts/aziot-compatibility.sh [hunk 5]
```diff
@@ -501,58 +547,75 @@ check_shared_library_dependency() {
...
 
-check_shared_library_dependency_display_util() {
+check_storage_space() {
+    storage_path=$1
+    application_size=$2
+    buffer=$3
+
+    # Check dependencies
+    if ! need_cmd df; then
+        wrap_warn "check_storage_space"
```
Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:575]: do we need two warning functions?

File: platform-validation/scripts/aziot-compatibility.sh [hunk 6]
```diff
@@ -595,11 +657,34 @@ aziotedge_check() {
...
+    eval binary_size='$'"$(echo "$ARCH"_iotedge_binaries_size)"
+    eval container_size='$'"$(echo "$ARCH"_iotedge_container_size)"
+    TOTAL_SIZE=$(echo $binary_size $container_size | awk '{print $1 + $2}')
+    if [ -z "$MOUNTPOINT" ]; then
+        wrap_debug "No Storage Path Provided, Using Present working directory"
+        MOUNTPOINT=$(pwd)
+    fi
+
+    check_storage_space "$MOUNTPOINT" "$TOTAL_SIZE" "$iot_edge_size_buffer"
+    ret="$?"
```
Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:675]: why assign and use the exit code? why not directly check for it

File: platform-validation/scripts/aziot-compatibility.sh [hunk 7]
```diff
@@ -595,11 +657,34 @@ aziotedge_check() {
...
+    elif [ $ARCH = armv7l ]; then
         SHARED_LIBRARIES="$(echo $SHARED_LIBRARIES "ld-linux-armhf.so.3")"
     fi
+
     check_shared_library_dependency
 
+    eval binary_size='$'"$(echo "$ARCH"_iotedge_binaries_size)"
+    eval container_size='$'"$(echo "$ARCH"_iotedge_container_size)"
+    TOTAL_SIZE=$(echo $binary_size $container_size | awk '{print $1 + $2}')
+    if [ -z "$MOUNTPOINT" ]; then
```
Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:669]: will MOUNTPOINT be a user input while running the application? then it should be part of the usage of the script?

File: platform-validation/scripts/aziot-compatibility.sh (hunk 5 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh:575]: One outputs a warning , one outputs a warning debug message. When we do the refactor of log levels, we can maybe give it a better name

File: platform-validation/scripts/aziot-compatibility.sh (hunk 7 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh:669]: It would be an Environment Variable input, not an input to the script.

File: platform-validation/scripts/aziot-compatibility.sh (hunk 6 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh:675]: because the next executed statement clears $?, hence saving it

File: platform-validation/scripts/aziot-compatibility.sh [hunk 8]
```diff
@@ -743,3 +744,8 @@ else
     fi
     "$APP_NAME"_check
 fi
+
+if ! echo "" | grep -E '^[0-9]+.?[0-9]+$'; then
```
Comment (damonbarry) [inline platform-validation/scripts/aziot-compatibility.sh:748]: Do you plan to leave this test in the script?

File: platform-validation/scripts/aziot-compatibility.sh [hunk 9]
```diff
@@ -501,58 +547,75 @@ check_shared_library_dependency() {
...
+    buffer=$3
+
+    # Check dependencies
+    if ! need_cmd df; then
+        wrap_warn "check_storage_space"
+        wrap_warning "Could not find df utility to calculate disk space, Skipping the check"
+        return
+    fi
+
+    if ! echo "$application_size" | grep -Eq '^[0-9]+.?[0-9]+$'; then
```
Comment (damonbarry) [inline platform-validation/scripts/aziot-compatibility.sh:580]: You should probably escape the '.' in the regex so it's interpreted as a period, not as "any character".

File: platform-validation/scripts/aziot-compatibility.sh [hunk 10]
```diff
@@ -595,11 +657,34 @@ aziotedge_check() {
         SHARED_LIBRARIES="$(echo $SHARED_LIBRARIES "ld-linux-x86-64.so.2")"
     elif [ $ARCH = aarch64 ]; then
         SHARED_LIBRARIES="$(echo $SHARED_LIBRARIES "ld-linux-aarch64.so.1")"
-    elif [ $ARCH = armv7 ]; then
+    elif [ $ARCH = armv7l ]; then
         SHARED_LIBRARIES="$(echo $SHARED_LIBRARIES "ld-linux-armhf.so.3")"
     fi
+
     check_shared_library_dependency
```
Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:665]: Can you add it as a separate function?

File: platform-validation/scripts/aziot-compatibility.sh [hunk 11]
```diff
@@ -501,58 +547,75 @@ check_shared_library_dependency() {
     # IOTEDGE_COMMON_SHARED_LIBRARIES is for both aziot-edged and aziot-identityd
```
Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:547]: can you remove this comment? as var name changed

File: platform-validation/scripts/aziot-compatibility.sh (hunk 8 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: Removed. Thanks for catching this!

File: platform-validation/scripts/aziot-compatibility.sh (hunk 9 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: Done

File: platform-validation/scripts/aziot-compatibility.sh (hunk 11 above)
Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: Done

File: platform-validation/scripts/aziot-compatibility.sh [hunk 12]
```diff
@@ -497,62 +543,77 @@ check_docker_api_version() {
...
 }
 
-check_shared_library_dependency_display_util() {
+check_storage_space() {
+    storage_path=$1
+    application_size=$2
+    buffer=$3
+
+    # Check dependencies
+    if ! need_cmd df; then
```
Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:572]: Did you check if this way of checking a command work? It doesn't work for me.

Comment (ggjjj) [inline platform-validation/scripts/aziot-compatibility.sh:572]: Did you check if this way of checking a command work? It doesn't work for me.

Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: Shell Check formatting broke it for me, I fixed it to the correct usage

Comment (nimanch) [reply platform-validation/scripts/aziot-compatibility.sh]: Shell Check formatting broke it for me, I fixed it to the correct usage
